Error BDD tests: "too many open files"

Running BDD tests locally results in this error:

<details><summary>expand</summary>

```
panic: error issuing command to docker-compose with args '[logs]':  pipe2: too many open files ()

goroutine 1 [running]:
github.com/hyperledger/aries-framework-go/test/bdd.runBddTests.func1.2()
        /home/llorllale/dev/hyperledger/aries-framework-go/test/bdd/bddtests_test.go:109 +0x23a
github.com/cucumber/godog.(*Suite).run(0xc000453680)
        /home/llorllale/go/pkg/mod/github.com/cucumber/godog@v0.8.1/suite.go:275 +0x190
github.com/cucumber/godog.(*runner).run(0xc00045ea80, 0x14)
        /home/llorllale/go/pkg/mod/github.com/cucumber/godog@v0.8.1/run.go:118 +0x25c
github.com/cucumber/godog.RunWithOptions(0x120db46, 0x6, 0x1373e78, 0x0, 0x1601d015ba71069c, 0x101, 0x120c312, 0x3, 0x120dc5a, 0x6, ...)
        /home/llorllale/go/pkg/mod/github.com/cucumber/godog@v0.8.1/run.go:215 +0xbc2
github.com/hyperledger/aries-framework-go/test/bdd.runBddTests(0x120c312, 0x3, 0x120dc5a, 0x6, 0x0)
        /home/llorllale/dev/hyperledger/aries-framework-go/test/bdd/bddtests_test.go:78 +0x2c9
github.com/hyperledger/aries-framework-go/test/bdd.TestMain(0xc000218880)
        /home/llorllale/dev/hyperledger/aries-framework-go/test/bdd/bddtests_test.go:68 +0x1c4
main.main()
        _testmain.go:82 +0x334
FAIL    github.com/hyperledger/aries-framework-go/test/bdd      17.533s
```
</details>

**Problem:** @soluchok pointed out that BDD tests are not doing proper cleanup. Some bdd test packages have started doing this cleanup individually (eg. `introduce`, `outofband`).

It would be great if this cleanup was done automatically using one of godog's hooks (eg. `AfterScenario`).

Things to watch out for:

* `BDDContext.AgentCtx` holds `context.Provider` which have no `Close()` method. Should be refactored to hold `framework.Aries` instead
* [Concurrency](https://github.com/cucumber/godog/issues/180)

--- a/test/bdd/agent/agent_controller_steps.go
+++ b/test/bdd/agent/agent_controller_steps.go
@@ -29,10 +29,13 @@ type ControllerSteps struct {
 }
 
 // NewControllerSteps creates steps for agent with controller
-func NewControllerSteps(ctx *context.BDDContext) *ControllerSteps {
-	return &ControllerSteps{
-		bddContext: ctx,
-	}
+func NewControllerSteps() *ControllerSteps {
+	return &ControllerSteps{}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (a *ControllerSteps) SetContext(ctx *context.BDDContext) {
+	a.bddContext = ctx
 }
 
 // RegisterSteps registers agent steps

--- a/test/bdd/agent/agent_sdk_steps.go
+++ b/test/bdd/agent/agent_sdk_steps.go
@@ -46,8 +46,8 @@ type SDKSteps struct {
 }
 
 // NewSDKSteps returns new agent from client SDK
-func NewSDKSteps(ctx *context.BDDContext) *SDKSteps {
-	return &SDKSteps{bddContext: ctx}
+func NewSDKSteps() *SDKSteps {
+	return &SDKSteps{}
 }
 
 func (a *SDKSteps) createAgent(agentID, inboundHost, inboundPort, scheme string) error {
@@ -219,10 +219,16 @@ func (a *SDKSteps) createFramework(agentID string, opts ...aries.Option) error {
 		return fmt.Errorf("failed to create context: %w", err)
 	}
 
+	a.bddContext.Agents[agentID] = agent
 	a.bddContext.AgentCtx[agentID] = ctx
 	a.bddContext.Messengers[agentID] = agent.Messenger()
 
 	return nil
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (a *SDKSteps) SetContext(ctx *context.BDDContext) {
+	a.bddContext = ctx
 }
 
 // RegisterSteps registers agent steps

--- a/test/bdd/bddtests_test.go
+++ b/test/bdd/bddtests_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/introduce"
 	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/issuecredential"
 	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/messaging"
+	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/outofband"
 	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/route"
 	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/verifiable"
 )
@@ -37,6 +38,14 @@ const (
 
 var composition []*dockerutil.Composition
 var composeFiles = []string{"./fixtures/agent-rest", "./fixtures/sidetree-mock"}
+
+// Feature of the aries framework under test.
+type feature interface {
+	// SetContext is called before every scenario is run with a fresh new context
+	SetContext(*bddctx.BDDContext)
+	// invoked once to register the steps on the suite
+	RegisterSteps(*godog.Suite)
+}
 
 func TestMain(m *testing.M) {
 	// default is to run all tests with tag @all
@@ -140,42 +149,43 @@ func generateUUID() string {
 }
 
 func FeatureContext(s *godog.Suite) {
-	bddContext, err := bddctx.NewBDDContext()
-	if err != nil {
-		panic(fmt.Sprintf("Error returned from NewBDDContext: %s", err))
+	features := features()
+
+	for _, f := range features {
+		f.RegisterSteps(s)
 	}
 
-	defer bddContext.Destroy()
+	var bddContext *bddctx.BDDContext
 
-	// set dynamic args
-	bddContext.Args[SideTreeURL] = "http://localhost:48326/document"
-	bddContext.Args[DIDDocPath] = "fixtures/sidetree-mock/config/didDocument.json"
+	s.BeforeScenario(func(interface{}) {
+		bddContext = bddctx.NewBDDContext()
+		// set dynamic args
+		bddContext.Args[SideTreeURL] = "http://localhost:48326/document"
+		bddContext.Args[DIDDocPath] = "fixtures/sidetree-mock/config/didDocument.json"
 
-	// Context is shared between tests
-	agent.NewSDKSteps(bddContext).RegisterSteps(s)
-	agent.NewControllerSteps(bddContext).RegisterSteps(s)
+		for _, f := range features {
+			f.SetContext(bddContext)
+		}
+	})
+	s.AfterScenario(func(_ interface{}, _ error) {
+		bddContext.Destroy()
+	})
+}
 
-	// Register did exchange tests
-	didexchange.NewDIDExchangeSDKSteps(bddContext).RegisterSteps(s)
-	didexchange.NewDIDExchangeControllerSteps(bddContext).RegisterSteps(s)
-
-	// Register introduce tests
-	introduce.NewIntroduceSDKSteps(bddContext).RegisterSteps(s)
-
-	// Register issuecredential tests
-	issuecredential.NewIssueCredentialSDKSteps(bddContext).RegisterSteps(s)
-
-	// Register did resolver tests
-	didresolver.NewDIDResolverSteps(bddContext).RegisterSteps(s)
-
-	// Register messaging tests
-	messaging.NewMessagingSDKSteps(bddContext).RegisterSteps(s)
-	messaging.NewMessagingControllerSteps(bddContext).RegisterSteps(s)
-
-	// Register router tests
-	route.NewRouteSDKSteps(bddContext).RegisterSteps(s)
-	route.NewRouteRESTSteps(bddContext).RegisterSteps(s)
-
-	// Register verifiable credential tests
-	verifiable.NewVerifiableCredentialSDKSteps(bddContext).RegisterSteps(s)
+func features() []feature {
+	return []feature{
+		agent.NewSDKSteps(),
+		agent.NewControllerSteps(),
+		didexchange.NewDIDExchangeSDKSteps(),
+		didexchange.NewDIDExchangeControllerSteps(),
+		introduce.NewIntroduceSDKSteps(),
+		issuecredential.NewIssueCredentialSDKSteps(),
+		didresolver.NewDIDResolverSteps(),
+		messaging.NewMessagingSDKSteps(),
+		messaging.NewMessagingControllerSteps(),
+		route.NewRouteSDKSteps(),
+		route.NewRouteRESTSteps(),
+		verifiable.NewVerifiableCredentialSDKSteps(),
+		outofband.NewOutOfBandSDKSteps(),
+	}
 }

--- a/test/bdd/features/outofband_e2e_sdk.feature
+++ b/test/bdd/features/outofband_e2e_sdk.feature
@@ -1,0 +1,21 @@
+#
+# Copyright SecureKey Technologies Inc. All Rights Reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Reference : https://github.com/hyperledger/aries-rfcs/blob/master/features/0434-outofband/README.md
+
+@all
+@outofband_e2e_sdk
+Feature: Out-Of-Band protocol
+
+  Background:
+    Given "Alice" agent is running on "localhost" port "random" with "http" as the transport provider
+    And "Bob" agent is running on "localhost" port "random" with "http" as the transport provider
+
+  Scenario: New connection after Alice sends an out-of-band request to Bob
+    Given "Alice" constructs an out-of-band request with no attachments
+    And "Alice" sends the request to "Bob" through an out-of-band channel
+    And "Bob" accepts the request and connects with "Alice"
+    Then "Alice" and "Bob" confirm their connection is "completed"

--- a/test/bdd/pkg/context/context.go
+++ b/test/bdd/pkg/context/context.go
@@ -17,6 +17,7 @@ import (
 	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
 	"github.com/hyperledger/aries-framework-go/pkg/didcomm/messaging/msghandler"
 	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
+	"github.com/hyperledger/aries-framework-go/pkg/framework/aries"
 	"github.com/hyperledger/aries-framework-go/pkg/framework/context"
 )
 
@@ -28,6 +29,7 @@ type BDDContext struct {
 	RouteClients       map[string]*route.Client
 	PublicDIDDocs      map[string]*did.Doc
 	PublicDIDs         map[string]string
+	Agents             map[string]*aries.Aries
 	AgentCtx           map[string]*context.Provider
 	MessageRegistrar   map[string]*msghandler.Registrar
 	Messengers         map[string]service.Messenger
@@ -39,12 +41,13 @@ type BDDContext struct {
 }
 
 // NewBDDContext create new BDDContext
-func NewBDDContext() (*BDDContext, error) {
-	instance := BDDContext{
+func NewBDDContext() *BDDContext {
+	return &BDDContext{
 		DIDExchangeClients: make(map[string]*didexchange.Client),
 		RouteClients:       make(map[string]*route.Client),
 		PublicDIDDocs:      make(map[string]*did.Doc),
 		PublicDIDs:         make(map[string]string),
+		Agents:             make(map[string]*aries.Aries),
 		AgentCtx:           make(map[string]*context.Provider),
 		MessageRegistrar:   make(map[string]*msghandler.Registrar),
 		Messengers:         make(map[string]service.Messenger),
@@ -53,16 +56,6 @@ func NewBDDContext() (*BDDContext, error) {
 		webhookURLs:        make(map[string]string),
 		webSocketConns:     make(map[string]*websocket.Conn),
 	}
-
-	return &instance, nil
-}
-
-// BeforeScenario execute code before bdd scenario
-func (b *BDDContext) BeforeScenario(scenarioOrScenarioOutline interface{}) {
-}
-
-// AfterScenario execute code after bdd scenario
-func (b *BDDContext) AfterScenario(interface{}, error) {
 }
 
 // Destroy BDD context
@@ -71,7 +64,13 @@ func (b *BDDContext) Destroy() {
 	for agentID, conn := range b.webSocketConns {
 		err := conn.Close(websocket.StatusNormalClosure, "bddtests destroy context")
 		if err != nil {
-			logger.Debugf("failed to close websocket connection for [%s] : %w", agentID, err)
+			logger.Warnf("failed to close websocket connection for [%s] : %w", agentID, err)
+		}
+	}
+
+	for _, agent := range b.Agents {
+		if err := agent.Close(); err != nil {
+			logger.Warnf("failed to teardown aries framework : %s", err.Error())
 		}
 	}
 }

--- a/test/bdd/pkg/didexchange/didexchange_controller_steps.go
+++ b/test/bdd/pkg/didexchange/didexchange_controller_steps.go
@@ -54,12 +54,16 @@ type ControllerSteps struct {
 }
 
 // NewDIDExchangeControllerSteps creates steps for didexchange with controller
-func NewDIDExchangeControllerSteps(ctx *context.BDDContext) *ControllerSteps {
+func NewDIDExchangeControllerSteps() *ControllerSteps {
 	return &ControllerSteps{
-		bddContext:    ctx,
 		invitations:   make(map[string]*didexchange.Invitation),
 		connectionIDs: make(map[string]string),
 	}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (a *ControllerSteps) SetContext(ctx *context.BDDContext) {
+	a.bddContext = ctx
 }
 
 // RegisterSteps registers agent steps

--- a/test/bdd/pkg/didexchange/didexchange_sdk_steps.go
+++ b/test/bdd/pkg/didexchange/didexchange_sdk_steps.go
@@ -32,9 +32,8 @@ type SDKSteps struct {
 }
 
 // NewDIDExchangeSDKSteps return new steps for didexchange using client SDK
-func NewDIDExchangeSDKSteps(ctx *context.BDDContext) *SDKSteps {
+func NewDIDExchangeSDKSteps() *SDKSteps {
 	return &SDKSteps{
-		bddContext:     ctx,
 		nextAction:     make(map[string]chan struct{}),
 		connectionID:   make(map[string]string),
 		invitations:    make(map[string]*didexchange.Invitation),
@@ -164,7 +163,8 @@ func (d *SDKSteps) WaitForPostEvent(agents, statesValue string) error {
 	return nil
 }
 
-func (d *SDKSteps) validateConnection(agentID, stateValue string) error {
+// ValidateConnection checks the agents connection status against the given value.
+func (d *SDKSteps) ValidateConnection(agentID, stateValue string) error {
 	conn, err := d.bddContext.DIDExchangeClients[agentID].GetConnection(d.connectionID[agentID])
 	if err != nil {
 		return fmt.Errorf("failed to query connection by id: %w", err)
@@ -333,7 +333,7 @@ func (d *SDKSteps) performDIDExchange(inviter, invitee string) error {
 			return err
 		}
 
-		err = d.validateConnection(agentID, completedStateValue)
+		err = d.ValidateConnection(agentID, completedStateValue)
 		if err != nil {
 			return err
 		}
@@ -394,6 +394,11 @@ func resolveDID(vdriRegistry vdriapi.Registry, did string, maxRetry int) (*diddo
 	return doc, err
 }
 
+// SetContext is called before every scenario is run with a fresh new context
+func (d *SDKSteps) SetContext(ctx *context.BDDContext) {
+	d.bddContext = ctx
+}
+
 // RegisterSteps registers did exchange steps
 func (d *SDKSteps) RegisterSteps(s *godog.Suite) { //nolint dupl
 	s.Step(`^"([^"]*)" creates invitation$`, d.createInvitation)
@@ -406,7 +411,7 @@ func (d *SDKSteps) RegisterSteps(s *godog.Suite) { //nolint dupl
 	s.Step(`^"([^"]*)" initiates connection with "([^"]*)" using public DID$`, d.createImplicitInvitationWithDID)
 	s.Step(`^"([^"]*)" waits for post state event "([^"]*)"$`, d.WaitForPostEvent)
 	s.Step(`^"([^"]*)" retrieves connection record and validates that connection state is "([^"]*)"$`,
-		d.validateConnection)
+		d.ValidateConnection)
 	s.Step(`^"([^"]*)" creates did exchange client$`, d.CreateDIDExchangeClient)
 	s.Step(`^"([^"]*)" approves did exchange request`, d.ApproveRequest)
 	s.Step(`^"([^"]*)" approves invitation request`, d.ApproveRequest)

--- a/test/bdd/pkg/didresolver/didresolver_steps.go
+++ b/test/bdd/pkg/didresolver/didresolver_steps.go
@@ -46,8 +46,13 @@ type Steps struct {
 }
 
 // NewDIDResolverSteps returns new steps for DID resolver tests
-func NewDIDResolverSteps(ctx *bddctx.BDDContext) *Steps {
-	return &Steps{bddContext: ctx}
+func NewDIDResolverSteps() *Steps {
+	return &Steps{}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (d *Steps) SetContext(ctx *bddctx.BDDContext) {
+	d.bddContext = ctx
 }
 
 func (d *Steps) createDIDDocument(agents, method string) error {

--- a/test/bdd/pkg/introduce/introduce_sdk_steps.go
+++ b/test/bdd/pkg/introduce/introduce_sdk_steps.go
@@ -36,13 +36,17 @@ type SDKSteps struct {
 }
 
 // NewIntroduceSDKSteps creates steps for introduce with SDK
-func NewIntroduceSDKSteps(ctx *context.BDDContext) *SDKSteps {
+func NewIntroduceSDKSteps() *SDKSteps {
 	return &SDKSteps{
-		bddContext: ctx,
-		clients:    make(map[string]*introduce.Client),
-		actions:    make(map[string]chan service.DIDCommAction),
-		events:     make(map[string]chan service.StateMsg),
+		clients: make(map[string]*introduce.Client),
+		actions: make(map[string]chan service.DIDCommAction),
+		events:  make(map[string]chan service.StateMsg),
 	}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (a *SDKSteps) SetContext(ctx *context.BDDContext) {
+	a.bddContext = ctx
 }
 
 // RegisterSteps registers agent steps
@@ -84,7 +88,8 @@ func (a *SDKSteps) createConnections(introducees, introducer string) error {
 	)
 
 	participants := introducees + "," + introducer
-	agentSDK := agent.NewSDKSteps(a.bddContext)
+	agentSDK := agent.NewSDKSteps()
+	agentSDK.SetContext(a.bddContext)
 
 	err := agentSDK.CreateAgentWithHTTPDIDResolver(participants, inboundHost, inboundPort, endpointURL, acceptDidMethod)
 	if err != nil {
@@ -95,7 +100,8 @@ func (a *SDKSteps) createConnections(introducees, introducer string) error {
 		return err
 	}
 
-	a.didExchangeSDKS = bddDIDExchange.NewDIDExchangeSDKSteps(a.bddContext)
+	a.didExchangeSDKS = bddDIDExchange.NewDIDExchangeSDKSteps()
+	a.didExchangeSDKS.SetContext(a.bddContext)
 
 	if err := a.didExchangeSDKS.WaitForPublicDID(participants, 10); err != nil {
 		return err
@@ -189,16 +195,6 @@ func (a *SDKSteps) checkHistoryEventsAndStop(agentID, events string) error {
 			}
 		case <-time.After(timeout):
 			return fmt.Errorf("waited for %s: history of events doesn't meet the expectation", stateID)
-		}
-	}
-
-	return a.stopServices()
-}
-
-func (a *SDKSteps) stopServices() error {
-	for _, ctx := range a.bddContext.AgentCtx {
-		if err := ctx.StorageProvider().Close(); err != nil {
-			return err
 		}
 	}
 

--- a/test/bdd/pkg/issuecredential/issuecredential_sdk_steps.go
+++ b/test/bdd/pkg/issuecredential/issuecredential_sdk_steps.go
@@ -61,13 +61,17 @@ type SDKSteps struct {
 }
 
 // NewIssueCredentialSDKSteps creates steps for the issuecredential with SDK
-func NewIssueCredentialSDKSteps(ctx *context.BDDContext) *SDKSteps {
+func NewIssueCredentialSDKSteps() *SDKSteps {
 	return &SDKSteps{
-		bddContext: ctx,
-		clients:    make(map[string]*issuecredential.Client),
-		actions:    make(map[string]chan service.DIDCommAction),
-		events:     make(map[string]chan service.StateMsg),
+		clients: make(map[string]*issuecredential.Client),
+		actions: make(map[string]chan service.DIDCommAction),
+		events:  make(map[string]chan service.StateMsg),
 	}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (a *SDKSteps) SetContext(ctx *context.BDDContext) {
+	a.bddContext = ctx
 }
 
 // RegisterSteps registers agent steps

--- a/test/bdd/pkg/messaging/messaging_controller_steps.go
+++ b/test/bdd/pkg/messaging/messaging_controller_steps.go
@@ -59,9 +59,8 @@ type ControllerSteps struct {
 }
 
 // NewMessagingControllerSteps return new steps for messaging using controller/REST binding
-func NewMessagingControllerSteps(ctx *context.BDDContext) *ControllerSteps {
+func NewMessagingControllerSteps() *ControllerSteps {
 	return &ControllerSteps{
-		bddContext:     ctx,
 		msgIDsBySender: make(map[string]string),
 	}
 }
@@ -529,6 +528,11 @@ func closeResponse(c io.Closer) {
 	if err != nil {
 		logger.Errorf("Failed to close response body : %s", err)
 	}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (d *ControllerSteps) SetContext(ctx *context.BDDContext) {
+	d.bddContext = ctx
 }
 
 // RegisterSteps registers messaging steps

--- a/test/bdd/pkg/messaging/messaging_sdk_steps.go
+++ b/test/bdd/pkg/messaging/messaging_sdk_steps.go
@@ -40,9 +40,8 @@ type SDKSteps struct {
 }
 
 // NewMessagingSDKSteps return new steps for messaging using client SDK
-func NewMessagingSDKSteps(ctx *context.BDDContext) *SDKSteps {
+func NewMessagingSDKSteps() *SDKSteps {
 	return &SDKSteps{
-		steps:           newMessagingSDKSteps(ctx),
 		genericMessages: make(map[string]*msgService),
 		basicMessages:   make(map[string]basic.Message),
 		msgIDsBySender:  make(map[string]string),
@@ -178,6 +177,13 @@ func (d *SDKSteps) receiveBasicMsg(agentID, msg, from string) error {
 	}
 
 	return nil
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (d *SDKSteps) SetContext(ctx *context.BDDContext) {
+	d.steps = &messagingSDKSteps{
+		bddContext: ctx,
+	}
 }
 
 // RegisterSteps registers messaging steps

--- a/test/bdd/pkg/messaging/messaging_steps.go
+++ b/test/bdd/pkg/messaging/messaging_steps.go
@@ -23,13 +23,6 @@ type messagingSDKSteps struct {
 	bddContext *context.BDDContext
 }
 
-// newMessagingSDKSteps return new steps for messaging using client SDK
-func newMessagingSDKSteps(ctx *context.BDDContext) *messagingSDKSteps {
-	return &messagingSDKSteps{
-		bddContext: ctx,
-	}
-}
-
 func (d *messagingSDKSteps) registerMsgService(agentID string, msgSvc dispatcher.MessageService) error {
 	registrar, ok := d.bddContext.MessageRegistrar[agentID]
 	if !ok {

--- a/test/bdd/pkg/outofband/outofband_sdk_steps.go
+++ b/test/bdd/pkg/outofband/outofband_sdk_steps.go
@@ -1,0 +1,188 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package outofband
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cucumber/godog"
+	"github.com/google/uuid"
+
+	"github.com/hyperledger/aries-framework-go/pkg/client/outofband"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
+	"github.com/hyperledger/aries-framework-go/test/bdd/pkg/context"
+	bddDIDExchange "github.com/hyperledger/aries-framework-go/test/bdd/pkg/didexchange"
+)
+
+// SDKSteps for the out-of-band protocol.
+type SDKSteps struct {
+	context         *context.BDDContext
+	oobClients      map[string]*outofband.Client
+	pendingRequests map[string]*outofband.Request
+	connectionIDs   map[string]string
+	bddDIDExchSDK   *bddDIDExchange.SDKSteps
+}
+
+// NewOutOfBandSDKSteps returns the out-of-band protocol's BDD steps using the SDK binding.
+func NewOutOfBandSDKSteps() *SDKSteps {
+	return &SDKSteps{
+		oobClients:      make(map[string]*outofband.Client),
+		pendingRequests: make(map[string]*outofband.Request),
+		connectionIDs:   make(map[string]string),
+		bddDIDExchSDK:   bddDIDExchange.NewDIDExchangeSDKSteps(),
+	}
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (sdk *SDKSteps) SetContext(ctx *context.BDDContext) {
+	sdk.context = ctx
+	sdk.bddDIDExchSDK.SetContext(ctx)
+}
+
+// RegisterSteps registers the BDD steps on the suite.
+func (sdk *SDKSteps) RegisterSteps(suite *godog.Suite) {
+	suite.Step(
+		`^"([^"]*)" constructs an out-of-band request with no attachments`, sdk.constructOOBRequestWithNoAttachments)
+	suite.Step(
+		`^"([^"]*)" sends the request to "([^"]*)" through an out-of-band channel`, sdk.sendRequestThruOOBChannel)
+	suite.Step(`^"([^"]*)" accepts the request and connects with "([^"]*)"`, sdk.acceptRequestAndConnect)
+	suite.Step(`^"([^"]*)" and "([^"]*)" confirm their connection is "([^"]*)"`, sdk.confirmConnections)
+}
+
+func (sdk *SDKSteps) constructOOBRequestWithNoAttachments(agentID string) error {
+	err := sdk.registerClients(agentID)
+	if err != nil {
+		return fmt.Errorf("failed to register outofband client : %w", err)
+	}
+
+	req, err := sdk.newRequest(agentID)
+	if err != nil {
+		return fmt.Errorf("failed to create an out-of-bound request : %w", err)
+	}
+
+	sdk.pendingRequests[agentID] = req
+
+	return nil
+}
+
+// sends a the sender's pending request to the receiver and returns the sender and receiver's new connection IDs.
+func (sdk *SDKSteps) sendRequestThruOOBChannel(senderID, receiverID string) error {
+	err := sdk.registerClients([]string{senderID, receiverID}...)
+	if err != nil {
+		return fmt.Errorf("failed to register framework clients : %w", err)
+	}
+
+	req, found := sdk.pendingRequests[senderID]
+	if !found {
+		return fmt.Errorf("no request found for %s", senderID)
+	}
+
+	delete(sdk.pendingRequests, senderID)
+
+	sdk.pendingRequests[receiverID] = req
+
+	return nil
+}
+
+func (sdk *SDKSteps) acceptRequestAndConnect(receiverID, senderID string) error {
+	request, found := sdk.pendingRequests[receiverID]
+	if !found {
+		return fmt.Errorf("no pending requests found for %s", receiverID)
+	}
+
+	delete(sdk.pendingRequests, receiverID)
+
+	receiver, found := sdk.oobClients[receiverID]
+	if !found {
+		return fmt.Errorf("no registered outofband client for %s", receiverID)
+	}
+
+	err := sdk.bddDIDExchSDK.RegisterPostMsgEvent(strings.Join([]string{senderID, receiverID}, ","), "completed")
+	if err != nil {
+		return fmt.Errorf("failed to register agents for didexchange post msg events : %w", err)
+	}
+
+	sdk.connectionIDs[receiverID], err = receiver.AcceptRequest(request)
+	if err != nil {
+		return fmt.Errorf("%s failed to accept out-of-band invitation : %w", receiverID, err)
+	}
+
+	err = sdk.bddDIDExchSDK.ApproveRequest(receiverID)
+	if err != nil {
+		return fmt.Errorf("failed to approve request for %s : %w", senderID, err)
+	}
+
+	err = sdk.bddDIDExchSDK.ApproveRequest(senderID)
+	if err != nil {
+		return fmt.Errorf("failed to approve request for %s : %w", senderID, err)
+	}
+
+	return nil
+}
+
+func (sdk *SDKSteps) confirmConnections(senderID, receiverID, status string) error {
+	err := sdk.bddDIDExchSDK.WaitForPostEvent(strings.Join([]string{senderID, receiverID}, ","), status)
+	if err != nil {
+		return fmt.Errorf("failed to wait for post events : %w", err)
+	}
+
+	err = sdk.bddDIDExchSDK.ValidateConnection(senderID, status)
+	if err != nil {
+		return fmt.Errorf("failed to validate connection status for %s : %w", senderID, err)
+	}
+
+	err = sdk.bddDIDExchSDK.ValidateConnection(receiverID, status)
+	if err != nil {
+		return fmt.Errorf("failed to validate connection status for %s : %w", receiverID, err)
+	}
+
+	return nil
+}
+
+func (sdk *SDKSteps) registerClients(agentIDs ...string) error {
+	for _, agent := range agentIDs {
+		if _, exists := sdk.oobClients[agent]; !exists {
+			client, err := outofband.New(sdk.context.AgentCtx[agent])
+			if err != nil {
+				return fmt.Errorf("failed to create new outofband client : %w", err)
+			}
+
+			sdk.oobClients[agent] = client
+		}
+
+		if _, exists := sdk.context.DIDExchangeClients[agent]; !exists {
+			err := sdk.bddDIDExchSDK.CreateDIDExchangeClient(agent)
+			if err != nil {
+				return fmt.Errorf("failed to create new didexchange client : %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (sdk *SDKSteps) newRequest(agentID string) (*outofband.Request, error) {
+	agent, found := sdk.oobClients[agentID]
+	if !found {
+		return nil, fmt.Errorf("no agent for %s was found", agentID)
+	}
+
+	req, err := agent.CreateRequest(outofband.WithAttachments(&decorator.Attachment{
+		ID:          uuid.New().String(),
+		Description: "dummy",
+		MimeType:    "text/plain",
+		Data: decorator.AttachmentData{
+			JSON: map[string]interface{}{},
+		},
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request for %s : %w", agentID, err)
+	}
+
+	return req, nil
+}

--- a/test/bdd/pkg/route/route_controller_steps.go
+++ b/test/bdd/pkg/route/route_controller_steps.go
@@ -34,10 +34,8 @@ type RESTSteps struct {
 }
 
 // NewRouteRESTSteps return steps for route using REST APIs
-func NewRouteRESTSteps(ctx *context.BDDContext) *RESTSteps {
-	return &RESTSteps{
-		bddContext: ctx,
-	}
+func NewRouteRESTSteps() *RESTSteps {
+	return &RESTSteps{}
 }
 
 // RegisterRoute registers the router for the agent.
@@ -99,6 +97,11 @@ func (d *RESTSteps) VerifyConnection(agentID, varName string) error {
 	}
 
 	return nil
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (d *RESTSteps) SetContext(ctx *context.BDDContext) {
+	d.bddContext = ctx
 }
 
 // RegisterSteps registers router steps

--- a/test/bdd/pkg/route/route_sdk_steps.go
+++ b/test/bdd/pkg/route/route_sdk_steps.go
@@ -21,10 +21,8 @@ type SDKSteps struct {
 }
 
 // NewRouteSDKSteps return steps for router using client SDK
-func NewRouteSDKSteps(ctx *context.BDDContext) *SDKSteps {
-	return &SDKSteps{
-		bddContext: ctx,
-	}
+func NewRouteSDKSteps() *SDKSteps {
+	return &SDKSteps{}
 }
 
 // CreateRouteClient creates route client
@@ -63,6 +61,11 @@ func (d *SDKSteps) VerifyConnection(agentID, varName string) error {
 	}
 
 	return nil
+}
+
+// SetContext is called before every scenario is run with a fresh new context
+func (d *SDKSteps) SetContext(ctx *context.BDDContext) {
+	d.bddContext = ctx
 }
 
 // RegisterSteps registers router steps

--- a/test/bdd/pkg/verifiable/verifiable_steps.go
+++ b/test/bdd/pkg/verifiable/verifiable_steps.go
@@ -30,10 +30,8 @@ type SDKSteps struct {
 }
 
 // NewVerifiableCredentialSDKSteps creates steps for verifiable credential with SDK
-func NewVerifiableCredentialSDKSteps(ctx *context.BDDContext) *SDKSteps {
-	return &SDKSteps{
-		bddContext: ctx,
-	}
+func NewVerifiableCredentialSDKSteps() *SDKSteps {
+	return &SDKSteps{}
 }
 
 const (
@@ -41,6 +39,11 @@ const (
 
 	jwsProof = "JWS"
 )
+
+// SetContext is called before every scenario is run with a fresh new context
+func (s *SDKSteps) SetContext(ctx *context.BDDContext) {
+	s.bddContext = ctx
+}
 
 // RegisterSteps registers Verifiable Credential steps.
 func (s *SDKSteps) RegisterSteps(gs *godog.Suite) {
@@ -169,7 +172,8 @@ func (s *SDKSteps) createDID(issuer, holder string) error {
 	)
 
 	participants := issuer + "," + holder
-	agentSDK := agent.NewSDKSteps(s.bddContext)
+	agentSDK := agent.NewSDKSteps()
+	agentSDK.SetContext(s.bddContext)
 
 	err := agentSDK.CreateAgentWithHTTPDIDResolver(participants, inboundHost, inboundPort, endpointURL, acceptDidMethod)
 	if err != nil {
@@ -180,7 +184,8 @@ func (s *SDKSteps) createDID(issuer, holder string) error {
 		return err
 	}
 
-	didExchangeSDK := bddDIDExchange.NewDIDExchangeSDKSteps(s.bddContext)
+	didExchangeSDK := bddDIDExchange.NewDIDExchangeSDKSteps()
+	didExchangeSDK.SetContext(s.bddContext)
 
 	return didExchangeSDK.WaitForPublicDID(participants, 10)
 }

